pkg/controlplane/controller/kubernetesservice/controller.go

```diff
@@ -169,10 +172,19 @@ func (c *Controller) UpdateKubernetesService(reconcile bool) error {
 		return err
 	}
 	endpointPorts := createEndpointPortSpec(c.PublicServicePort, "https")
-	if err := c.EndpointReconciler.ReconcileEndpoints(kubernetesServiceName, c.PublicIP, endpointPorts, reconcile); err != nil {
-		return err
+
+	var code int
+	var err error
+	result := c.restClient.Get().AbsPath("/readyz").Do(context.TODO())
+	result.StatusCode(&code)
+	// add the endpoint if it is ready, otherwise remove it
+	if result.Error() == nil && code == http.StatusOK {
+		err = c.EndpointReconciler.ReconcileEndpoints(kubernetesServiceName, c.PublicIP, endpointPorts, reconcile)
+	} else {
+		err = c.EndpointReconciler.RemoveEndpoints(kubernetesServiceName, c.PublicIP, endpointPorts)
```

Comment:
i feel that more nuances may be needed to make a decision to remove endpoints. The most important one that i can tell is fail open. The reconciler must not remove endpoints if the action leads to 0 endpoints.

Reply:
> The usecase in point is that certain users create large amount of secrets and as an operator i cannot ask them not to. The solution should make sure that such users don't see a cascading error because removing the k8s endpoint will cause all traffic to skew to remaining servers(usually 1 or 2 in my case).

if you have a kms errors as explained in the bug all the apiserver will become unhealthy, I do not understand the special case of "remove endpoints if unhealthy but leave at least one unhealthy",

Reply:
We can implement it any way that helps a fail open scenario. Im not specifically proposing leaving just one, only proposing that if all endpoints are unhealthy, then fail open and reconcile all.

Reply:
I don't like to implement complex heuristics in something as critical as this, one apiserver can be good or bad, but not good depending if all as bad, ... that doesn't look something easy to think about

Reply:
I agree about keeping simple....but service discovery in most tools comply with this heuristic. I feel the change could be a KEP. This change in itself could cause unavailability.

Reply:
> but service discovery in most tools comply with this heuristic of fail open.

do you have some examples, just curiosity

Reply:
Bahavior of NLB: https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html
```
 If all targets fail health checks at the same time in all enabled Availability Zones, the load balancer fails open. The effect of the fail open is to allow traffic to all targets in all enabled Availability Zones, regardless of their health status.
```

Envoy Panic routing https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold
```
However, if the percentage of available hosts in the cluster becomes too low, Envoy will disregard health status and balance either amongst all hosts or no hosts. This is known as the panic threshold. The panic threshold is used to avoid a situation in which host failures cascade throughout the cluster as load increases.
```

Reply:
@jyotimahapatra I see, but this is not the same, this is not a loadbalancer, this is a mechanism each apiserver has to publish its own endpoint, so the Service implementation (this is what you are referring to in your links), usually kube-proxy chooses the backend. 
Another thing is what you are proposing will not work here because if all are unhealthy, and all of them try to remove their endpoint, once they detect there are no endpoint they will try to add the endpoint, and then because they see they are not the only one, they will try to remove it again ...

What we can do and I think is a good idea, is to use the Endpoint Ready and NotReady addresses instead of removing the address, this will allow the Service implementation to implement the heuristics that you mean

You have David's blessing as api-machinery chair, add me to the kep approver we'll flesh out the details there

Reply:
> so the Service implementation (this is what you are referring to in your links), usually kube-proxy chooses the backend.

Correct. Hence the server should expose the endpoints as is and let client decide. Removing the endpoint in the server does not allow client to make any decisions.

> use the Endpoint Ready and NotReady addresses instead of removing the address, this will allow the Service implementation to implement the heuristics that you mean

That sounds very reasonable. There is still a problem at kube-proxy end, from my recollection of kube-proxy's implementation , i think kube-proxy does not fail open either and will remove both endpoints. I'll verify whether my statement is accurate

Reply:
but that is out of scope of the KEP, this reconciler provides the API and the semantics and the configuration options, the implementations options are a non goal